### PR TITLE
feat: add DialOption for IAM DB Authentication

### PR DESCRIPTION
**Pull request by kurtisvg**
## Change Description

Adds the WithDialIAMAuthN DialOption, for dialing an instance with an IAMAuthN option different than the Dialer setting.

### CODE REVIEW

File: dialer_test.go [hunk 1]
```diff
@@ -295,3 +299,143 @@ func TestDialerUserAgent(t *testing.T) {
...
 	}
 }
+
+func TestWarmup(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	stop := mock.StartServerProxy(t, inst)
+	defer func() {
```
Comment (enocom) [inline dialer_test.go:309]: nit: `defer stop()`

File: dialer_test.go [hunk 2]
```diff
@@ -295,3 +299,143 @@ func TestDialerUserAgent(t *testing.T) {
...
+	}
+}
+
+func TestDialDialerOptsConflicts(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	stop := mock.StartServerProxy(t, inst)
+	defer func() {
```
Comment (enocom) [inline dialer_test.go:379]: `defer stop()`

File: dialer.go [hunk 3]
```diff
@@ -297,24 +301,33 @@ func (d *Dialer) Close() error {
 	return nil
 }
 
-func (d *Dialer) instance(connName string) (*cloudsql.Instance, error) {
+// instance is a helper function for returning the appropriate instance object in a threadsafe way.
+// It will create a new instance object, modify the existing one, or leave it unchanged as needed.
+func (d *Dialer) instance(connName string, rCfg *cloudsql.RefreshCfg) (*cloudsql.Instance, error) {
```
Comment (enocom) [inline dialer.go:306]: nit: `rCfg` is clear as just `r`. The type tells the rest.

File: dialer.go [hunk 4]
```diff
@@ -297,24 +301,33 @@ func (d *Dialer) Close() error {
...
 	// Check instance cache
 	d.lock.RLock()
 	i, ok := d.instances[connName]
 	d.lock.RUnlock()
-	if !ok {
+	// Check if the instance exists and that the refresh cfg is the same
+	if !ok || (rCfg != nil && *rCfg != i.RefreshCfg) {
```
Comment (enocom) [inline dialer.go:312]: nit: `(rCfg != nil && *rCfg != i.RefreshCfg)` might be clearer as a function declared inline.

``` golang
newConfig := func(r *cloudsql.RefreshCfg) bool { return r != nil && *r != i.RefreshCfg }
if !ok || newConfig(rCfg) { /* ... */ }
```

File: internal/cloudsql/instance.go [hunk 5]
```diff
@@ -148,6 +154,7 @@ func NewInstance(
 	refreshTimeout time.Duration,
 	ts oauth2.TokenSource,
 	dialerID string,
+	rCfg RefreshCfg,
```
Comment (enocom) [inline internal/cloudsql/instance.go:157]: `r` is enough.

File: internal/cloudsql/instance.go [hunk 6]
```diff
@@ -213,6 +221,19 @@ func (i *Instance) InstanceEngineVersion(ctx context.Context) (string, error) {
 	return res.md.version, nil
 }
 
+func (i *Instance) UpdateRefresh(cfg RefreshCfg) {
+	i.resultGuard.Lock()
```
Comment (enocom) [inline internal/cloudsql/instance.go:225]:
``` golang
i.resultGuard.Lock()
defer i.resultGuard.Unlock()
/* ... */
```

**Review by enocom — COMMENTED** [review]
A few minor things

File: dialer.go (hunk 4 above)
Comment (kurtisvg) [reply dialer.go:312]: Adding a function seems to obfuscate for me. I updated the comment here instead.

File: dialer_test.go [hunk 7]
```diff
@@ -295,3 +299,143 @@ func TestDialerUserAgent(t *testing.T) {
...
+			}
+			d.sqladmin = svc
+			defer func() {
+				if err := cleanup(); err != nil {
+					t.Fatalf("%v", err)
+				}
+			}()
+
+			// Dial once with the "default" options
+			testSuccessfulDial(t, d, ctx, "my-project:my-region:my-instance", test.warmupOpts...)
```
[the remark this answers is not in the document]
Comment (kurtisvg) [reply dialer_test.go]: Good catch.

File: dialer.go [hunk 8]
```diff
@@ -297,24 +301,33 @@ func (d *Dialer) Close() error {
...
-func (d *Dialer) instance(connName string) (*cloudsql.Instance, error) {
+// instance is a helper function for returning the appropriate instance object in a threadsafe way.
+// It will create a new instance object, modify the existing one, or leave it unchanged as needed.
+func (d *Dialer) instance(connName string, r *cloudsql.RefreshCfg) (*cloudsql.Instance, error) {
 	// Check instance cache
 	d.lock.RLock()
 	i, ok := d.instances[connName]
 	d.lock.RUnlock()
-	if !ok {
+	// If the instance hasn't been creted yet or if the needed refreshCfg has changed
```
Comment (enocom) [inline dialer.go:311]: delete "needed"?

**Review by enocom — APPROVED** [review]

**Pull request merged**